### PR TITLE
a8n: Introduce "automation" GraphQL schema

### PR DESCRIPTION
**Pull request by tsenart**
This commit introduces an initial GraphQL schema for the a8n product.

To help reviewing, it makes the necessary changes to allow us to use
GraphQL schema extension, rather than bundling everything in a single
file.

Once we discuss, iterate and agree on this schema, we'll merge it and proceed to implement it, using both [prototype code](https://github.com/sourcegraph/sourcegraph/tree/a8n/campaigns) and new code as appropriate.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -212,7 +212,6 @@ require (
 
 replace (
 	github.com/google/zoekt => github.com/sourcegraph/zoekt v0.0.0-20190821144538-4e4eef546d3e
-	github.com/graph-gophers/graphql-go => github.com/sourcegraph/graphql-go v0.0.0-20180929065141-c790ffc3c46a
```
Comment (tsenart) [inline go.mod]: @sqs: I couldn't find the reason for us depending on our fork, and we needed the support for `extends` that landed more recently upstream, hence this change. Did I miss something obvious?

**Comment by tsenart** [comment]
As the build indicates, it seems our GraphQL library mandates that we have resolvers implemented for all the new stuff.

Anyone can think of an easy way to make the build pass without having to create all of this scaffold? If not, I'll copy the interfaces defined in the prototype branch over.

File: go.mod (hunk 1 above)
Comment (felixfbecker) [reply go.mod]: This was the commit it pointed to: https://github.com/sourcegraph/graphql-go/commit/c790ffc3c46ad5c1ad44d506ac89e5d0112e7079

Looks like it was to include the fix of https://github.com/graph-gophers/graphql-go/commit/021755791b94a00ca421743bf71b0e130f0a8aec, which is now also upstream master

**Comment by felixfbecker** [comment]
@tsenart do you want to merge this before we have  a (real) implementation for the resolvers? If not I think it's fine if the build fails for now (that way we know that we are still missing implementations)

File: shared/gulpfile.ts [hunk 2]
```diff
@@ -20,7 +21,7 @@ export async function watchGraphQLTypes(): Promise<void> {
 
 /** Generates the TypeScript types for the GraphQL schema */
 export async function graphQLTypes(): Promise<void> {
-    const schemaStr = await readFile(GRAPHQL_SCHEMA_PATH, 'utf8')
+    const schemaStr = (await Promise.all(fg.sync(GRAPHQL_SCHEMA_PATH).map(f => readFile(f)))).join('')
```
Comment (felixfbecker) [inline shared/gulpfile.ts]:
```suggestion
    const schemaStr = (await Promise.all(glob(GRAPHQL_SCHEMA_PATH).map(f => readFile(f)))).join('\n')
```

File: shared/gulpfile.ts [hunk 3]
```diff
@@ -9,8 +9,9 @@ import { readFile, writeFile } from 'mz/fs'
 import path from 'path'
 import { format, resolveConfig } from 'prettier'
 import { draftV7resolver } from './draftV7Resolver'
+import fg from 'fast-glob'
```
Comment (felixfbecker) [inline shared/gulpfile.ts]:
```suggestion
import glob from 'fast-glob'
```

File: go.mod (hunk 1 above)
Comment (sqs) [reply go.mod]: Last time I tried upgrading this library to upstream master (a few months ago), there were test failures. If it passes tests then I don’t know of any other reasons why we weren’t on the latest (beyond what Felix said).

**Comment by tsenart** [comment]
> @tsenart do you want to merge this before we have a (real) implementation for the resolvers?

I wanted to, yes. I think that makes it easier to incrementally grasp the domain and the existing prototype code, rather than all at once (which is really hard for me to do).

File: shared/gulpfile.ts [hunk 4]
```diff
@@ -20,7 +21,8 @@ export async function watchGraphQLTypes(): Promise<void> {
 
 /** Generates the TypeScript types for the GraphQL schema */
 export async function graphQLTypes(): Promise<void> {
-    const schemaStr = await readFile(GRAPHQL_SCHEMA_PATH, 'utf8')
+    const files = await glob<string>(GRAPHQL_SCHEMA_PATH)
+    const schemaStr = (await Promise.all(files.map(file => readFile(file)))).join('\n')
```
Comment (tsenart) [inline shared/gulpfile.ts]: @felixfbecker: Just for me to seize this learning opportunity: This code is now fully asynchronous. What are the risks of mixing sync and async code in Typescript like I was doing before?

**Comment by keegancsmith** [comment]
> > @tsenart do you want to merge this before we have a (real) implementation for the resolvers?
> 
> I wanted to, yes. I think that makes it easier to incrementally grasp the domain and the existing prototype code, rather than all at once (which is really hard for me to do).

Maybe put the changes under mutation and query into an experimental type? Then put stub implementations that return errors, which you later flesh out. That way we can ship this in master to customers, without any issues. Then when we are ready to promote this to ready, we can move the resolvers from experimental into mutation/query.

**Comment by tsenart** [comment]
> Maybe put the changes under mutation and query into an experimental type? 

What about all the other changes to existing types?

**Comment by keegancsmith** [comment]
From our last chat I believe the plan going forward is to rather merge in the full a8n branch than piecemeal it out. As such this PR may be abandoned. If I am wrong, please add me back as a reviewer.

**Pull request closed without merging**